refactor: remove parent, parenttype and parentfield columns from non-child table doctypes

phot0n opened the pull request:
This pr aims to remove parent, parenttype, parentfield and idx columns from non-child table doctypes

idea/goal is to make is child table checkbox work in a similar way as is tree checkbox which adds fields to the doctype - but at schema/table structure level as opposed to adding fields in json/docfield

TODO:
- [ ] find more references (?)
- [x] add tests


will (hopefully) close: #15228 

previous pr https://github.com/frappe/frappe/pull/15410 got messed up hence opening a new one :P

ankush commented:
do you need to add these fields in doctype.json? Can't they just be handled like the way they were before but conditionally in case of child table doc? 🤔 

The reason why I am asking is you'll only fix it for frappe there are many apps out there who already use this 😅

phot0n commented:
ah yes, i was a bit confused regarding that like how will the apps other than frappe work - will they have these fields in their metas (doctype.json) whereas only frappe will have them via schema?
hence, in order to make everything consistent i just did it via meta 😅 
Though i didn't think of the existing apps 🤔 
I'll try it that way - will reach out if i get stuck 👍

phot0n committed:
refactor: remove parent, parenttype, parentfield, idx columns from non-child table doctypes

* feat: add parent, parenttype, idx, parentfield columns to doctypes when transitioning from normal -> child table

* fix: remove parent, parenttype, parentfield, idx from DocType DocType

phot0n commented:
@ankush it kinda seem to work lol - please have a look https://github.com/phot0n/frappe/commit/5bfef626318203f1529721c822613b386b957802

ankush commented:
@phot0n 

Ideally, these 4 fields should be added as DB columns and dealt with **without** adding it in `.json` schema files. Is this how it will work now? If so first undo changes to all JSON files.

phot0n commented:
closing for now. will continue later

phot0n committed:
chore(notifications.py): remove unused function get_unread_emails

phot0n committed:
chore: remove patch

phot0n committed:
test: child table creation and transitioning

phot0n committed:
chore(notifications.py): remove unused get_unseen_likes

phot0n committed:
chore: adding back idx column

This is needed for ordering in display of links as it keeps
track of the clicks on a particular link item

phot0n committed:
fix: use int for docstatus in get_default_df

phot0n committed:
fix: add index on parent in child table doctype

gavindsouza commented:
Was this completed before merging? I see pending todos and a refactor without tests! Reminded me of something I read recently:

_"refactors without test coverage is oxymoron"_

**Edit: Seems like a test was added, but the description hasn't been updated. I'm assuming it covers all the changes made.**

cc: @phot0n @rmehta

rmehta commented:
Nothing kills motivation and productivity than delay. Since this was in “develop”, we took some risks.

The code is internal enough to be called in 100s of teats. 

> On 07-Feb-2022, at 7:27 PM, gavin ***@***.***> wrote:
> 
> ﻿
> Was this completed before merging? I see pending todos and a refactor without tests! Reminded me of something I read recently:
> 
> "refactors without test coverage is oxymoron"
> 
> cc: @phot0n @rmehta
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.
